### PR TITLE
Partition aware policy

### PR DESCRIPTION
**Pull request by HarshDaryani896**
Added PartitionAwarePolicy and Tests for it.

### CODE REVIEW

File: example_batch_test.go [hunk 1]
```diff
@@ -3,8 +3,9 @@ package gocql_test
 import (
 	"context"
 	"fmt"
-	"github.com/gocql/gocql"
```
Comment (ashetkar) [inline example_batch_test.go:6]: You can replace it in-place rather than as a separate import line. Same for other files.

File: example_lwt_batch_test.go [hunk 2]
```diff
@@ -39,12 +40,12 @@ func ExampleSession_MapExecuteBatchCAS() {
 	executeBatch := func(ck2Version int) {
 		b := session.NewBatch(gocql.LoggedBatch)
 		b.Entries = append(b.Entries, gocql.BatchEntry{
-			Stmt:       "UPDATE my_lwt_batch_table SET value=? WHERE pk=? AND ck=? IF version=?",
```
Comment (ashetkar) [inline example_lwt_batch_test.go:42]: Revert unnecessary changes.

File: example_nulls_test.go [hunk 3]
```diff
@@ -28,7 +29,7 @@ func Example_nulls() {
 	scanner := session.Query(`SELECT id, value FROM stringvals`).Iter().Scanner()
 	for scanner.Next() {
 		var (
-			id int32
+			id  int32
```
Comment (ashetkar) [inline example_nulls_test.go:32]: Revert unnecessary change.

File: example_paging_test.go [hunk 4]
```diff
@@ -38,7 +39,7 @@ func Example_paging() {
 		scanner := iter.Scanner()
 		for scanner.Next() {
 			var (
-				id int
+				id          int
```
Comment (ashetkar) [inline example_paging_test.go:42]: Revert unnecessary change.

File: example_set_test.go [hunk 5]
```diff
@@ -45,7 +46,7 @@ func Example_set() {
 	scanner := session.Query(`SELECT id, value FROM sets`).Iter().Scanner()
 	for scanner.Next() {
 		var (
-			id int32
+			id  int32
```
Comment (ashetkar) [inline example_set_test.go:49]: Remove unnecessary change

File: example_udt_struct_test.go [hunk 6]
```diff
@@ -3,13 +3,14 @@ package gocql_test
...
-	"github.com/gocql/gocql"
 	"log"
+
+	"github.com/yugabyte/gocql"
 )
 
 type MyUDT struct {
 	FieldA string `cql:"field_a"`
-	FieldB int32 `cql:"field_b"`
+	FieldB int32  `cql:"field_b"`
```
Comment (ashetkar) [inline example_udt_struct_test.go:13]: Revert unnecessary change here and all other places.

File: host_source.go [hunk 7]
```diff
@@ -392,6 +392,8 @@ type ringDescriber struct {
 	mu              sync.Mutex
 	prevHosts       []*HostInfo
 	prevPartitioner string
+
+	currHostsyb []*HostInfo
```
Comment (ashetkar) [inline host_source.go:396]: Suggestion: `currHostsyb` -> `currYbHosts`

File: host_source.go [hunk 8]
```diff
@@ -574,14 +576,35 @@ func (r *ringDescriber) GetHosts() ([]*HostInfo, string, error) {
...
+
+	for _, k := range r.currHostsyb {
+		if k.ConnectAddress().Equal(ip) {
+			host = k
+			break
+		}
+	}
+
+	if host == nil {
+		return nil, errors.New("host not found in peers table")
```
Comment (ashetkar) [inline host_source.go:603]: "... not found in system.peers table"

File: marshal.go [hunk 9]
```diff
@@ -161,6 +161,77 @@ func Marshal(info TypeInfo, value interface{}) ([]byte, error) {
 	return nil, fmt.Errorf("can not marshal %T into %s", value, info)
 }
 
+func Marshalyb(info TypeInfo, value interface{}) ([]byte, error) {
```
Comment (ashetkar) [inline marshal.go:164]: Suggestion: `Marshalyb` -> `MarshalYb` or `YbMarshal`

File: marshal.go [hunk 10]
```diff
@@ -1269,6 +1340,35 @@ func marshalTimestamp(info TypeInfo, value interface{}) ([]byte, error) {
 	return nil, marshalErrorf("can not marshal %T into %s", value, info)
 }
 
+func marshalTimestampyb(info TypeInfo, value interface{}) ([]byte, error) {
```
Comment (ashetkar) [inline marshal.go:1343]: Suggestion: `marshalTimestampyb` -> `marshalTimestampYb` or `ybMarshalTimestamp`

File: policies.go [hunk 11]
```diff
@@ -620,6 +620,244 @@ func (t *tokenAwareHostPolicy) Pick(qry ExecutableQuery) NextHost {
...
+		meta.resetTokenRing(p.partitioner, p.hosts.get(), p.logger)
+		p.updateReplicas(meta, p.getKeyspaceName())
+		p.metadata.Store(meta)
+	}
+	p.mu.Unlock()
+
+	p.fallback.RemoveHost(host)
+}
+
+func (p *ybPartitionAwareHostPolicy) SetPartitioner(partitioner string) {
```
Comment (ashetkar) [inline policies.go:742]: Is this required for our policy? Please remove, if not. Similarly, check other functions.

File: policies_test.go [hunk 12]
```diff
@@ -7,13 +7,318 @@ package gocql
...
+	}
+	content, err := ioutil.ReadAll(res.Body)
+	res.Body.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+	return string(content)
+}
+
+func localReadandWrite(s string, t *testing.T) (int, int) {
```
Comment (ashetkar) [inline policies_test.go:35]: `localReadandWrite` -> `localReadAndWrite`

File: query_executor.go [hunk 13]
```diff
@@ -12,7 +12,9 @@ type ExecutableQuery interface {
 	retryPolicy() RetryPolicy
 	speculativeExecutionPolicy() SpeculativeExecutionPolicy
 	GetRoutingKey() ([]byte, error)
+	GetRoutingKeyyb() ([]byte, error)
```
Comment (ashetkar) [inline query_executor.go:15]: Suggestions:
`GetRoutingKeyyb()` -> `GetRoutingKeyYb()` or `GetYbRoutingKey()`
`KeyspaceAndTableyb()` -> `KeyspaceAndTableYb()` or `YbKeyspaceAndTable()`

File: session.go [hunk 14]
```diff
@@ -1067,6 +1067,7 @@ func (q *Query) retryPolicy() RetryPolicy {
 
 // Keyspace returns the keyspace the query will be executed against.
 func (q *Query) Keyspace() string {
+
```
Comment (ashetkar) [inline session.go:1070]: Remove unnecessary change

**Review by ashetkar — CHANGES_REQUESTED** [review]
I'm still reviewing the PR. These comments are mostly around formatting/naming

File: example_batch_test.go (hunk 1 above)
Comment (HarshDaryani896) [reply example_batch_test.go:6]: I did only change in-place but when I save the file VS-CODE itself does this change.

**Review by ashetkar — COMMENTED** [review]
Few more comments/questions

File: policies.go (hunk 11 above)
Comment (HarshDaryani896) [reply policies.go:670]: All policies extend\return [HostSelectionPolicy](https://github.com/yugabyte/gocql/blob/partition_aware_policy/policies.go#L265) interface so each policy needs to have a implementation of all the functions that are part of the interface due to which we cannot remove such functions but since we do not need these functions for PartitionAwarePolicy they are empty implementations for YBPartitionAwarePolicy.

**Review by nchandrappa — APPROVED** [review]
